user-contracts-mon/internal/monitor.go

```diff
@@ -0,0 +1,146 @@
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/NicoNex/echotron/v3"
+	"github.com/threefoldtech/tfgrid-sdk-go/grid-client/deployer"
+	"github.com/threefoldtech/tfgrid-sdk-go/grid-client/graphql"
+)
+
+// Monitor struct of parsed configration
+type Monitor struct {
+	Bot      echotron.API
+	Mnemonic string
+	BotToken string
+	Network  string
+	interval int
+}
+
+// NewMonitor creates a new monitor from parsed config/env file
+func NewMonitor(conf Config) (Monitor, error) {
+	mon := Monitor{}
+	mon.Mnemonic = conf.mnemonic
+	mon.Network = conf.network
+	mon.interval = conf.interval
+	mon.BotToken = conf.botToken
+	mon.Bot = echotron.NewAPI(mon.BotToken)
+
+	return mon, nil
+}
+
+// StartMonitoring starts monitoring the contracts with
+// specific mnemonics and notify them every fixed interval
+func (mon Monitor) StartMonitoring(tfPluginClient deployer.TFPluginClient, chatID int64) {
+	ticker := time.NewTicker(time.Duration(mon.interval) * time.Hour)
+	for ; true; <-ticker.C {
+
+		contractsInGracePeriod, err := getContractsInGracePeriod(tfPluginClient)
+		if err != nil {
+			_, err := mon.Bot.SendMessage("Failed to get contracts in grace period", chatID, nil)
+			if err != nil {
+				log.Println(err)
+			}
+
+			return
+		}
+
+		if contractsInGracePeriod != "" {
+			_, err = mon.Bot.SendMessage(contractsInGracePeriod, chatID, nil)
+			if err != nil {
+				return
+			}
+		}
+
+		contractsAgainstDownNodes, err := getContractsAgainstDownNodes(tfPluginClient)
+		if err != nil {
+			_, err = mon.Bot.SendMessage("Failed to get contracts against down nodes", chatID, nil)
+			if err != nil {
+				log.Println(err)
+			}
+			return
+		}
+
+		if contractsAgainstDownNodes != "" {
+			_, err = mon.Bot.SendMessage(contractsAgainstDownNodes, chatID, nil)
+			if err != nil {
+				return
+			}
+		}
+	}
+}
+
+func getContractsInGracePeriod(tfPluginClient deployer.TFPluginClient) (string, error) {
+	contracts, err := tfPluginClient.ContractsGetter.ListContractsByTwinID([]string{"GracePeriod"})
+	if err != nil {
+		return "", err
+	}
+
+	info := ""
+	for _, contract := range contracts.NameContracts {
+		info += fmt.Sprintf("- %s\n", contract.ContractID)
+	}
+
+	for _, contract := range contracts.NodeContracts {
+		info += fmt.Sprintf("- %s\n", contract.ContractID)
+	}
+
+	for _, contract := range contracts.RentContracts {
+		info += fmt.Sprintf("- %s\n", contract.ContractID)
+	}
+	if info == "" {
+		return "", nil
+	}
+	return "contracts in grace period:\n" + info, nil
+}
+
+func getContractsAgainstDownNodes(tfPluginClient deployer.TFPluginClient) (string, error) {
+	contracts, err := tfPluginClient.ContractsGetter.ListContractsByTwinID([]string{"Created"})
+	if err != nil {
+		return "", err
+	}
+
+	nodeContracts := contracts.NodeContracts
+	nodeContracts = append(nodeContracts, contracts.RentContracts...)
+
+	contractsIds := ""
+	upNodes := make(chan string)
+
+	for _, contract := range nodeContracts {
+		go isNodeUp(tfPluginClient, contract, upNodes)
```

Comment:
do we need to add `contract := contract` to not let it use the same contract?

Reply:
can you explain more what you mean?

Reply:
there is a problem with using loop variables with goroutines: https://go.dev/blog/loopvar-preview
but your implementation is correct